File: LeetCodePractice/src/DFS/EvaluateDivision.java
package DFS;

import java.util.HashMap;
import java.util.HashSet;
import java.util.Map;
import java.util.Set;

/*
Equations are given in the format A / B = k, where A and B are variables represented as strings, and k is a real number (floating point number). Given some queries, return the answers. If the answer does not exist, return -1.0.

Example:
Given a / b = 2.0, b / c = 3.0. 
queries are: a / c = ?, b / a = ?, a / e = ?, a / a = ?, x / x = ? . 
return [6.0, 0.5, -1.0, 1.0, -1.0 ].

The input is: vector<pair<string, string>> equations, vector<double>& values, vector<pair<string, string>> queries , where equations.size() == values.size(), and the values are positive. This represents the equations. Return vector<double>.

According to the example above:

equations = [ ["a", "b"], ["b", "c"] ],
values = [2.0, 3.0],
queries = [ ["a", "c"], ["b", "a"], ["a", "e"], ["a", "a"], ["x", "x"] ]. 
The input is always valid. You may assume that evaluating the queries will result in no division by zero and there is no contradiction.
*/
public class EvaluateDivision {
    Map<String, Map<String, Double>> graph;
    Set<String> visited;

    public double[] calcEquation(String[][] equations, double[] values, String[][] queries) {
        graph = new HashMap<>();
        for (int i = 0; i < equations.length; i++) {
            String[] eq = equations[i];
            String u = eq[0];
            String v = eq[1];
            graph.putIfAbsent(u, new HashMap<>());
            graph.get(u).put(v, values[i]);
            graph.putIfAbsent(v, new HashMap<>());
            graph.get(v).put(u, 1.0 / values[i]);
        }
        double[] res = new double[queries.length];
        for (int i = 0; i < queries.length; i++) {
            String[] query = queries[i];
            String start = query[0];
            String end = query[1];
            res[i] = dfs(start, end, new HashSet<>());
        }
        return res;
    }

    private double dfs(String start, String end, Set<String> visited) {
        if (visited.contains(start)) {
            return -1.0;
        }
        if (!graph.containsKey(start)) {
            return -1.0;
        }
        if (graph.get(start).containsKey(end)) {
            return graph.get(start).get(end);
        }
        visited.add(start);
        Map<String, Double> neighbors = graph.get(start);
        double res = -1.0;
        for (String nb : neighbors.keySet()) {
            double nb_val = graph.get(start).get(nb);
            double next = dfs(nb, end, visited);
            if (next != -1.0) {
                res = nb_val * next;
                break;
            }
        }
        visited.remove(start);
        return res;
    }
}

File: LeetCodePython/EvaluateDivision.py
"""
You are given an array of variable pairs equations and an array of real numbers values, where equations[i] = [Ai, Bi] and values[i] represent the equation Ai / Bi = values[i]. Each Ai or Bi is a string that represents a single variable.

You are also given some queries, where queries[j] = [Cj, Dj] represents the jth query where you must find the answer for Cj / Dj = ?.

Return the answers to all queries. If a single answer cannot be determined, return -1.0.

Note: The input is always valid. You may assume that evaluating the queries will not result in division by zero and that there is no contradiction.



Example 1:

Input: equations = [["a","b"],["b","c"]], values = [2.0,3.0], queries = [["a","c"],["b","a"],["a","e"],["a","a"],["x","x"]]
Output: [6.00000,0.50000,-1.00000,1.00000,-1.00000]
Explanation:
Given: a / b = 2.0, b / c = 3.0
queries are: a / c = ?, b / a = ?, a / e = ?, a / a = ?, x / x = ?
return: [6.0, 0.5, -1.0, 1.0, -1.0 ]
Example 2:

Input: equations = [["a","b"],["b","c"],["bc","cd"]], values = [1.5,2.5,5.0], queries = [["a","c"],["c","b"],["bc","cd"],["cd","bc"]]
Output: [3.75000,0.40000,5.00000,0.20000]
Example 3:

Input: equations = [["a","b"]], values = [0.5], queries = [["a","b"],["b","a"],["a","c"],["x","y"]]
Output: [0.50000,2.00000,-1.00000,-1.00000]


Constraints:

1 <= equations.length <= 20
equations[i].length == 2
1 <= Ai.length, Bi.length <= 5
values.length == equations.length
0.0 < values[i] <= 20.0
1 <= queries.length <= 20
queries[i].length == 2
1 <= Cj.length, Dj.length <= 5
Ai, Bi, Cj, Dj consist of lower case English letters and digits.
"""
from collections import deque
class EvaluateDivision(object):
    def calcEquation(self, equations, values, queries):
        """
        :type equations: List[List[str]]
        :type values: List[float]
        :type queries: List[List[str]]/
        :rtype: List[float]
        """
        graph = {}
        def add_edge(x, y, value):
            if x in graph:
                graph[x].append((y, value))
            else:
                graph[x] = [(y, value)]

        for ([x, y], value) in zip(equations, values):
            add_edge(x, y, value)
            add_edge(y, x, 1 / value)

        def find_prod(s, e):
            if s not in graph or e not in graph:
                return -1.0
            if s == e:
                return 1.0
            q = deque([(s, 1.0)])
            visited = {s}
            while q:
                cur, prod = q.popleft()
                if cur == e:
                    return prod
                for nb, val in graph[cur]:
                    if nb not in visited:
                        visited.add(nb)
                        q.append((nb, prod * val))
            return -1.0

        return [find_prod(s, e) for s, e in queries]